### PR TITLE
feat(cardinal): event queueing and events package. 

### PR DESCRIPTION
**Pull request by pyrofolium**
Closes: #425
## What is the purpose of the change

Added Event queueing to the hub with events getting broadcasted on calling the Flush method. 

Additionally moved Events.go from Server to an events package. 

Created a test_utils package on the side for shared functionality among tests. 

## Brief Changelog

See above.

events are processed vertically. Meaning one thread per connection operating on the entire event queue during the flush. Each event in the queue is processed serially per connection.  
 

## Testing

moved the events related tests from server_test to events package which necessitated the creation of the test_utils package.

### CODE REVIEW

File: cardinal/events/events.go [hunk 1]
```diff
@@ -47,34 +65,43 @@ func (h *EventHub) Run() {
...
-					if err != nil {
-						unregisterConnection(conn)
-						log.Logger.Error().Err(err)
+					for _, event := range eh.eventQueue {
+						err := conn.SetWriteDeadline(time.Now().Add(5 * time.Second))
+						if err != nil {
+							go func() {
+								eh.UnregisterConnection(conn)
+							}()
+							log.Logger.Error().Err(err)
```
Comment (pyrofolium) [inline cardinal/events/events.go:87]: Have to unregister on the main thread.

File: cardinal/events/events.go [hunk 2]
```diff
@@ -23,21 +23,39 @@ func CreateEventHub() *EventHub {
...
+	Flush                chan bool
 	Unregister           chan *websocket.Conn
 	Register             chan *websocket.Conn
 	Shutdown             chan bool
+	eventQueue           []*Event
+}
+
+func (eh *EventHub) BroadcastEvent(event *Event) {
+	eh.Broadcast <- event
+}
```
[the remark this answers is not in the document]
Comment (technicallyty) [reply cardinal/events/events.go:41]: assuming this is the method users will use, if not ok to leave as is :-)

File: cardinal/events/events.go [hunk 3]
```diff
@@ -23,21 +23,39 @@ func CreateEventHub() *EventHub {
...
 }
 
 type EventHub struct {
 	websocketConnections map[*websocket.Conn]bool
-	Broadcast            chan []byte
+	Broadcast            chan *Event
+	Flush                chan bool
 	Unregister           chan *websocket.Conn
 	Register             chan *websocket.Conn
 	Shutdown             chan bool
```
Comment (technicallyty) [inline cardinal/events/events.go:35]: do all these fields need to be exported?

**Review by technicallyty — APPROVED** [review]
LGTM - do we have any PR lined up for showing how this works in systems?

**Review by namrapatel — APPROVED** [review]
Seems fine but I'm curious how we're gonna hook this up to actually work with Nakama and send notifications to clients, I'm assuming there are more PR(s) to come on that

**Comment by pyrofolium** [comment]
> LGTM - do we have any PR lined up for showing how this works in systems?

Yes, right now it's just this separate unused module. The next PR integrates everything.

**Comment by pyrofolium** [comment]
> Seems fine but I'm curious how we're gonna hook this up to actually work with Nakama and send notifications to clients, I'm assuming there are more PR(s) to come on that

Yes there's a PR that has to forward all of this to nakama. But that's in starter-game-template, and that one involves a full update to starter-game-template, or not... if someone updates starter-game-template before I get to it.

**Comment by namrapatel** [comment]
> > Seems fine but I'm curious how we're gonna hook this up to actually work with Nakama and send notifications to clients, I'm assuming there are more PR(s) to come on that
> 
> Yes there's a PR that has to forward all of this to nakama. But that's in starter-game-template, and that one involves a full update to starter-game-template, or not... if someone updates starter-game-template before I get to it.

Yeah that sounds good, I personally don't think we can consider an Events System complete until it is demonstrated to work with Nakama. I'll try to get `starter-game-template` caught up by then (I assigned myself a ticket for it) but ofc no promises :(

**Pull request merged**